File: packages/crypto/src/seed.ts
/**
 * BIP-39 Mnemonic Seed Generation + HD Key Derivation for Auth Box.
 *
 * This module implements the "Unstoppable" layer -- a seed phrase that can
 * deterministically derive all keys needed by Auth Box:
 *
 *   seed phrase (24 words, BIP-39)
 *     → master key (PBKDF2-HMAC-SHA512)
 *       → m/auth_box'/vault'/0'    → vault encryption key
 *       → m/auth_box'/sync'/0'     → sync encryption key
 *       → m/auth_box'/agent'/0'    → agent delegation root key
 *       → m/auth_box'/agent'/<n>'  → per-agent sub-key (revocable)
 *       → m/auth_box'/derive'/<n>' → deterministic password derivation
 *
 * Key properties:
 *   - Seed phrase is the ONLY recovery mechanism (no server needed)
 *   - All keys are deterministically derived (same seed = same keys)
 *   - Agent sub-keys can be individually revoked without affecting others
 *   - Deterministic passwords eliminate the need to store passwords at all
 *
 * Dependencies: @noble/hashes (SHA-256, HMAC-SHA512, PBKDF2)
 */

import { sha256 } from '@noble/hashes/sha256';
import { hmac } from '@noble/hashes/hmac';
import { sha512 } from '@noble/hashes/sha512';
import { pbkdf2 } from '@noble/hashes/pbkdf2';

// ─── BIP-39 English Wordlist (2048 words) ─────────────────────────────────
// We embed a minimal loader -- the full wordlist is loaded lazily.
// For now, we use the standard English BIP-39 wordlist.

let _wordlist: string[] | null = null;

/**
 * Set the BIP-39 wordlist. Must be called before generateMnemonic/validateMnemonic.
 * Typically loaded from a separate file to keep the core bundle small.
 */
export function setWordlist(words: string[]): void {
  if (words.length !== 2048) {
    throw new Error('BIP-39 wordlist must contain exactly 2048 words');
  }
  _wordlist = words;
}

function getWordlist(): string[] {
  if (!_wordlist) {
    throw new Error('BIP-39 wordlist not loaded. Call setWordlist() first.');
  }
  return _wordlist;
}

// ─── BIP-39 Mnemonic Generation ───────────────────────────────────────────

/**
 * Generate a BIP-39 mnemonic phrase.
 *
 * @param strength - Entropy bits: 128 (12 words), 160 (15), 192 (18), 224 (21), 256 (24 words)
 * @returns Space-separated mnemonic words
 */
export function generateMnemonic(strength: 128 | 160 | 192 | 224 | 256 = 256): string {
  const wordlist = getWordlist();

  // 1. Generate random entropy
  const entropy = new Uint8Array(strength / 8);
  crypto.getRandomValues(entropy);

  // 2. Calculate checksum: first (strength/32) bits of SHA-256(entropy)
  const hash = sha256(entropy);
  const checksumBits = strength / 32;

  // 3. Combine entropy + checksum into bit string
  let bits = '';
  for (const byte of entropy) {
    bits += byte.toString(2).padStart(8, '0');
  }
  for (let i = 0; i < checksumBits; i++) {
    bits += ((hash[Math.floor(i / 8)] >> (7 - (i % 8))) & 1).toString();
  }

  // 4. Split into 11-bit groups → word indices
  const words: string[] = [];
  for (let i = 0; i < bits.length; i += 11) {
    const index = parseInt(bits.slice(i, i + 11), 2);
    words.push(wordlist[index]);
  }

  return words.join(' ');
}

/**
 * Validate a BIP-39 mnemonic phrase.
 *
 * @returns true if the mnemonic has valid word count, all words are in the wordlist,
 *          and the checksum is correct.
 */
export function validateMnemonic(mnemonic: string): boolean {
  const wordlist = getWordlist();
  const words = mnemonic.trim().split(/\s+/);

  // Valid word counts: 12, 15, 18, 21, 24
  const validCounts = [12, 15, 18, 21, 24];
  if (!validCounts.includes(words.length)) return false;

  // Convert words to bit string
  let bits = '';
  for (const word of words) {
    const index = wordlist.indexOf(word);
    if (index === -1) return false;
    bits += index.toString(2).padStart(11, '0');
  }

  // Split entropy and checksum
  const strength = (words.length * 11 * 32) / 33;
  const entropyBits = bits.slice(0, strength);
  const checksumBits = bits.slice(strength);

  // Reconstruct entropy bytes
  const entropy = new Uint8Array(strength / 8);
  for (let i = 0; i < entropy.length; i++) {
    entropy[i] = parseInt(entropyBits.slice(i * 8, (i + 1) * 8), 2);
  }

  // Verify checksum
  const hash = sha256(entropy);
  const expectedChecksumLength = strength / 32;
  let expectedChecksum = '';
  for (let i = 0; i < expectedChecksumLength; i++) {
    expectedChecksum += ((hash[Math.floor(i / 8)] >> (7 - (i % 8))) & 1).toString();
  }

  return checksumBits === expectedChecksum;
}

// ─── BIP-39 Mnemonic → Seed ───────────────────────────────────────────────

/**
 * Convert a mnemonic phrase to a 512-bit seed using PBKDF2-HMAC-SHA512.
 *
 * @param mnemonic - BIP-39 mnemonic phrase
 * @param passphrase - Optional passphrase (default: empty string per BIP-39 spec)
 * @returns 64-byte seed
 */
export function mnemonicToSeed(mnemonic: string, passphrase: string = ''): Uint8Array {
  const encoder = new TextEncoder();
  const mnemonicBytes = encoder.encode(mnemonic.normalize('NFKD'));
  const salt = encoder.encode('mnemonic' + passphrase.normalize('NFKD'));

  return pbkdf2(sha512, mnemonicBytes, salt, {
    c: 2048,  // BIP-39 specifies 2048 iterations
    dkLen: 64, // 512-bit output
  });
}

// ─── HD Key Derivation (BIP-32 style, Auth Box paths) ─────────────────────

const HARDENED_OFFSET = 0x80000000;

// Auth Box custom derivation purpose (registered in our namespace)
const AUTH_BOX_PURPOSE = 0x41425800; // "ABX\0" as uint32

export enum KeyPurpose {
  VAULT = 0,    // m/ABX'/vault'/0'  → vault encryption key
  SYNC = 1,     // m/ABX'/sync'/0'   → sync encryption key
  AGENT = 2,    // m/ABX'/agent'/n'  → agent delegation keys
  DERIVE = 3,   // m/ABX'/derive'/n' → deterministic password derivation
  AUTH = 4,     // m/ABX'/auth'/0'   → authentication signing key
}

interface HDKey {
  key: Uint8Array;     // 32-byte private key material
  chainCode: Uint8Array; // 32-byte chain code for child derivation
}

/**
 * Derive the HD master key from a BIP-39 seed.
 * Uses HMAC-SHA512 with "authbox seed" as the key (different from Bitcoin's "Bitcoin seed").
 */
function masterKeyFromSeed(seed: Uint8Array): HDKey {
  const I = hmac(sha512, new TextEncoder().encode('authbox seed'), seed);
  return {
    key: I.slice(0, 32),
    chainCode: I.slice(32, 64),
  };
}

/**
 * Derive a hardened child key at the given index.
 * Only hardened derivation is used (index | 0x80000000) for security.
 */
function deriveHardenedChild(parent: HDKey, index: number): HDKey {
  const data = new Uint8Array(37);
  data[0] = 0x00; // hardened: 0x00 || key || index
  data.set(parent.key, 1);
  const indexBytes = new DataView(data.buffer);
  indexBytes.setUint32(33, index | HARDENED_OFFSET, false);

  const I = hmac(sha512, parent.chainCode, data);
  return {
    key: I.slice(0, 32),
    chainCode: I.slice(32, 64),
  };
}

/**
 * Derive a key for a specific Auth Box purpose from the seed.
 *
 * Path: m / AUTH_BOX_PURPOSE' / purpose' / index'
 *
 * @param seed - 64-byte BIP-39 seed
 * @param purpose - Key purpose (vault, sync, agent, derive, auth)
 * @param index - Sub-index (e.g., agent number, site index)
 * @returns 32-byte derived key
 */
export function deriveKey(seed: Uint8Array, purpose: KeyPurpose, index: number = 0): Uint8Array {
  const master = masterKeyFromSeed(seed);
  const purposeNode = deriveHardenedChild(master, AUTH_BOX_PURPOSE);
  const typeNode = deriveHardenedChild(purposeNode, purpose);
  const leaf = deriveHardenedChild(typeNode, index);
  return leaf.key;
}

// ─── Deterministic Password Derivation ────────────────────────────────────

const PASSWORD_CHARS = {
  lower: 'abcdefghijklmnopqrstuvwxyz',
  upper: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
  digits: '0123456789',
  symbols: '!@#$%^&*()-_=+[]{}|;:,.<>?',
};

export interface DerivePasswordOptions {
  length?: number;     // Default: 20
  lowercase?: boolean; // Default: true
  uppercase?: boolean; // Default: true
  digits?: boolean;    // Default: true
  symbols?: boolean;   // Default: true
  counter?: number;    // Default: 0 (increment to generate new password for same site)
}

/**
 * Deterministically derive a password for a specific site from the seed.
 *
 * This is the most radical innovation: if the user has their seed phrase,
 * they can regenerate the password for any site WITHOUT a vault.
 *
 * Path: m / ABX' / derive' / siteHash' → HKDF → password characters
 *
 * @param seed - 64-byte BIP-39 seed
 * @param site - Site identifier (domain name, e.g., "github.com")
 * @param options - Password generation options
 * @returns Deterministic password string
 */
export function derivePassword(
  seed: Uint8Array,
  site: string,
  options: DerivePasswordOptions = {},
): string {
  const {
    length = 20,
    lowercase = true,
    uppercase = true,
    digits = true,
    symbols = true,
    counter = 0,
  } = options;

  // Build character set
  let charset = '';
  if (lowercase) charset += PASSWORD_CHARS.lower;
  if (uppercase) charset += PASSWORD_CHARS.upper;
  if (digits) charset += PASSWORD_CHARS.digits;
  if (symbols) charset += PASSWORD_CHARS.symbols;

  if (charset.length === 0) {
    throw new Error('At least one character class must be enabled');
  }

  // Derive site-specific key material
  const encoder = new TextEncoder();
  const siteHash = sha256(encoder.encode(site.toLowerCase() + '\0' + counter.toString()));
  const siteIndex = new DataView(siteHash.buffer).getUint32(0, false);

  const siteKey = deriveKey(seed, KeyPurpose.DERIVE, siteIndex);

  // Expand key material to enough bytes for password generation
  // Use HMAC-SHA512 for expansion (deterministic, uniform)
  const normalizedSite = site.toLowerCase();
  const expanded = hmac(sha512, siteKey, encoder.encode('password:' + normalizedSite + ':' + counter));

  // Convert to password characters (uniform distribution via rejection sampling)
  const password: string[] = [];
  let byteIdx = 0;

  while (password.length < length) {
    if (byteIdx >= expanded.length) {
      // Need more bytes -- rehash
      const more = hmac(sha512, siteKey, new Uint8Array([...expanded, ...encoder.encode(password.length.toString())]));
      for (let i = 0; i < more.length && password.length < length; i++) {
        const charIdx = more[i] % charset.length;
        password.push(charset[charIdx]);
      }
      break;
    }

    const charIdx = expanded[byteIdx] % charset.length;
    password.push(charset[charIdx]);
    byteIdx++;
  }

  return password.join('');
}

// ─── Convenience Types ────────────────────────────────────────────────────

export interface SeedKeyBundle {
  vaultKey: Uint8Array;   // For encrypting the vault
  syncKey: Uint8Array;    // For encrypting sync payloads
  authKey: Uint8Array;    // For authentication signing
  agentRootKey: Uint8Array; // Root key for agent delegation
}

/**
 * Derive all primary keys from a seed in one call.
 */
export function deriveAllKeys(seed: Uint8Array): SeedKeyBundle {
  return {
    vaultKey: deriveKey(seed, KeyPurpose.VAULT, 0),
    syncKey: deriveKey(seed, KeyPurpose.SYNC, 0),
    authKey: deriveKey(seed, KeyPurpose.AUTH, 0),
    agentRootKey: deriveKey(seed, KeyPurpose.AGENT, 0),
  };
}

/**
 * Derive a per-agent delegation key.
 * Each agent gets its own key that can be independently revoked.
 */
export function deriveAgentKey(seed: Uint8Array, agentIndex: number): Uint8Array {
  return deriveKey(seed, KeyPurpose.AGENT, agentIndex);
}

File: apps/ios/AuthBoxCrypto/Sources/AuthBoxCrypto/Seed.swift
import CryptoKit
import Foundation

/// BIP-39 mnemonic generation + HD key derivation for Auth Box.
///
/// Mirrors `@authbox/crypto` seed.ts exactly. All constants and algorithms
/// are identical to ensure cross-platform seed recovery compatibility.
///
/// Derivation path: m / ABX' / purpose' / index'
/// Master key HMAC domain: "authbox seed"
public enum Seed {

    // MARK: - Constants (must match TypeScript exactly)

    private static let hardenedOffset: UInt32 = 0x80000000
    private static let authBoxPurpose: UInt32 = 0x41425800  // "ABX\0" as uint32

    private static let passwordCharsLower   = "abcdefghijklmnopqrstuvwxyz"
    private static let passwordCharsUpper   = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    private static let passwordCharsDigits  = "0123456789"
    private static let passwordCharsSymbols = "!@#$%^&*()-_=+[]{}|;:,.<>?"

    // MARK: - BIP-39 Mnemonic Generation

    /// Generate a BIP-39 mnemonic phrase.
    ///
    /// - Parameter strength: Entropy bits (128=12 words, 256=24 words). Default: 256
    /// - Returns: Space-separated mnemonic words
    public static func generateMnemonic(strength: Int = 256) -> String {
        let wordlist = WordlistEN.words
        precondition([128, 160, 192, 224, 256].contains(strength))

        // 1. Generate random entropy
        let entropyBytes = strength / 8
        var entropy = Data(count: entropyBytes)
        entropy.withUnsafeMutableBytes { ptr in
            _ = SecRandomCopyBytes(kSecRandomDefault, entropyBytes, ptr.baseAddress!)
        }

        // 2. Checksum: first (strength/32) bits of SHA-256(entropy)
        let hash = Data(SHA256.hash(data: entropy))
        let checksumBits = strength / 32

        // 3. Build bit string: entropy + checksum
        var bits = ""
        for byte in entropy {
            bits += String(byte, radix: 2).leftPadded(toLength: 8, with: "0")
        }
        for i in 0..<checksumBits {
            let byteIndex = i / 8
            let bitIndex = 7 - (i % 8)
            bits += String((hash[byteIndex] >> bitIndex) & 1)
        }

        // 4. Split into 11-bit groups -> word indices
        var words: [String] = []
        for i in stride(from: 0, to: bits.count, by: 11) {
            let start = bits.index(bits.startIndex, offsetBy: i)
            let end = bits.index(start, offsetBy: 11)
            let index = Int(bits[start..<end], radix: 2)!
            words.append(wordlist[index])
        }

        return words.joined(separator: " ")
    }

    /// Validate a BIP-39 mnemonic phrase.
    ///
    /// Checks word count, wordlist membership, and checksum.
    public static func validateMnemonic(_ mnemonic: String) -> Bool {
        let wordlist = WordlistEN.words
        let words = mnemonic.trimmingCharacters(in: .whitespaces)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }

        let validCounts = [12, 15, 18, 21, 24]
        guard validCounts.contains(words.count) else { return false }

        // Convert words to bit string
        var bits = ""
        for word in words {
            guard let index = wordlist.firstIndex(of: word) else { return false }
            bits += String(index, radix: 2).leftPadded(toLength: 11, with: "0")
        }

        // Split entropy and checksum
        let strength = (words.count * 11 * 32) / 33
        let entropyBits = String(bits.prefix(strength))
        let checksumBits = String(bits.suffix(bits.count - strength))

        // Reconstruct entropy bytes
        var entropy = Data(count: strength / 8)
        for i in 0..<entropy.count {
            let start = entropyBits.index(entropyBits.startIndex, offsetBy: i * 8)
            let end = entropyBits.index(start, offsetBy: 8)
            entropy[i] = UInt8(entropyBits[start..<end], radix: 2)!
        }

        // Verify checksum
        let hash = Data(SHA256.hash(data: entropy))
        let expectedChecksumLength = strength / 32
        var expectedChecksum = ""
        for i in 0..<expectedChecksumLength {
            let byteIndex = i / 8
            let bitIndex = 7 - (i % 8)
            expectedChecksum += String((hash[byteIndex] >> bitIndex) & 1)
        }

        return checksumBits == expectedChecksum
    }

    // MARK: - BIP-39 Mnemonic -> Seed

    /// Convert a mnemonic to a 512-bit seed using PBKDF2-HMAC-SHA512.
    ///
    /// Parameters match BIP-39 spec exactly:
    /// - 2048 iterations
    /// - Salt: "mnemonic" + passphrase (NFKD normalized)
    /// - 64-byte output
    public static func mnemonicToSeed(_ mnemonic: String, passphrase: String = "") -> Data {
        let mnemonicData = Data(mnemonic.precomposedStringWithCompatibilityMapping.utf8)
        let saltString = "mnemonic" + passphrase.precomposedStringWithCompatibilityMapping
        let saltData = Data(saltString.utf8)

        // PBKDF2-HMAC-SHA512, 2048 iterations, 64-byte output
        var derivedKey = Data(count: 64)
        _ = derivedKey.withUnsafeMutableBytes { derivedPtr in
            mnemonicData.withUnsafeBytes { passwordPtr in
                saltData.withUnsafeBytes { saltPtr in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordPtr.baseAddress?.assumingMemoryBound(to: Int8.self),
                        mnemonicData.count,
                        saltPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        saltData.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA512),
                        2048,
                        derivedPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                        64
                    )
                }
            }
        }

        return derivedKey
    }

    // MARK: - HD Key Derivation

    private struct HDKey {
        let key: Data       // 32-byte private key material
        let chainCode: Data // 32-byte chain code
    }

    /// Derive master key from BIP-39 seed.
    /// HMAC-SHA512 with "authbox seed" as key (NOT "Bitcoin seed").
    private static func masterKeyFromSeed(_ seed: Data) -> HDKey {
        let hmacKey = SymmetricKey(data: Data("authbox seed".utf8))
        let I = Data(HMAC<SHA512>.authenticationCode(for: seed, using: hmacKey))
        return HDKey(
            key: I.prefix(32),
            chainCode: I.suffix(from: 32)
        )
    }

    /// Derive a hardened child key.
    /// Data format: 0x00 || parent.key || uint32BE(index | 0x80000000)
    private static func deriveHardenedChild(_ parent: HDKey, index: UInt32) -> HDKey {
        var data = Data(count: 37)
        data[0] = 0x00
        data.replaceSubrange(1..<33, with: parent.key)
        let hardenedIndex = index | hardenedOffset
        data[33] = UInt8((hardenedIndex >> 24) & 0xFF)
        data[34] = UInt8((hardenedIndex >> 16) & 0xFF)
        data[35] = UInt8((hardenedIndex >> 8) & 0xFF)
        data[36] = UInt8(hardenedIndex & 0xFF)

        let hmacKey = SymmetricKey(data: parent.chainCode)
        let I = Data(HMAC<SHA512>.authenticationCode(for: data, using: hmacKey))
        return HDKey(
            key: I.prefix(32),
            chainCode: I.suffix(from: 32)
        )
    }

    /// Derive a key for a specific Auth Box purpose.
    ///
    /// Path: m / AUTH_BOX_PURPOSE' / purpose' / index'
    ///
    /// - Parameters:
    ///   - seed: 64-byte BIP-39 seed
    ///   - purpose: Key purpose (vault, sync, agent, derive, auth)
    ///   - index: Sub-index (e.g., agent number)
    /// - Returns: 32-byte derived key
    public static func deriveKey(seed: Data, purpose: KeyPurpose, index: UInt32 = 0) -> Data {
        let master = masterKeyFromSeed(seed)
        let purposeNode = deriveHardenedChild(master, index: authBoxPurpose)
        let typeNode = deriveHardenedChild(purposeNode, index: purpose.rawValue)
        let leaf = deriveHardenedChild(typeNode, index: index)
        return leaf.key
    }

    /// Derive all primary keys from a seed.
    public static func deriveAllKeys(seed: Data) -> SeedKeyBundle {
        SeedKeyBundle(
            vaultKey: deriveKey(seed: seed, purpose: .vault, index: 0),
            syncKey: deriveKey(seed: seed, purpose: .sync, index: 0),
            authKey: deriveKey(seed: seed, purpose: .auth, index: 0),
            agentRootKey: deriveKey(seed: seed, purpose: .agent, index: 0)
        )
    }

    /// Derive a per-agent delegation key.
    public static func deriveAgentKey(seed: Data, agentIndex: UInt32) -> Data {
        deriveKey(seed: seed, purpose: .agent, index: agentIndex)
    }

    // MARK: - Deterministic Password Derivation

    /// Deterministically derive a password for a site from the seed.
    ///
    /// Same seed + site = same password, every time, offline.
    /// Mirrors the TypeScript implementation exactly for cross-platform compatibility.
    ///
    /// Algorithm:
    /// 1. siteHash = SHA-256(site.lower() + "\0" + counter)
    /// 2. siteIndex = first 4 bytes of siteHash as uint32 big-endian
    /// 3. siteKey = deriveKey(seed, .derive, siteIndex)
    /// 4. expanded = HMAC-SHA512(siteKey, "password:" + site.lower() + ":" + counter)
    /// 5. password chars selected by expanded[i] % charset.length
    public static func derivePassword(
        seed: Data,
        site: String,
        options: DerivePasswordOptions = DerivePasswordOptions()
    ) -> String {
        let length = options.length
        let counter = options.counter

        // Build character set
        var charset = ""
        if options.lowercase { charset += passwordCharsLower }
        if options.uppercase { charset += passwordCharsUpper }
        if options.digits    { charset += passwordCharsDigits }
        if options.symbols   { charset += passwordCharsSymbols }

        precondition(!charset.isEmpty, "At least one character class must be enabled")
        let charsetArray = Array(charset)

        // Derive site-specific key material
        let normalizedSite = site.lowercased()
        let siteHashInput = Data((normalizedSite + "\0" + String(counter)).utf8)
        let siteHash = Data(SHA256.hash(data: siteHashInput))

        let siteIndex: UInt32 = siteHash.withUnsafeBytes { ptr in
            let b = ptr.bindMemory(to: UInt8.self)
            return (UInt32(b[0]) << 24) | (UInt32(b[1]) << 16) | (UInt32(b[2]) << 8) | UInt32(b[3])
        }

        let siteKey = deriveKey(seed: seed, purpose: .derive, index: siteIndex)

        // Expand key material using HMAC-SHA512
        let expandInput = Data(("password:" + normalizedSite + ":" + String(counter)).utf8)
        let hmacKey = SymmetricKey(data: siteKey)
        let expanded = Data(HMAC<SHA512>.authenticationCode(for: expandInput, using: hmacKey))

        // Convert to password characters
        var password: [Character] = []
        var byteIdx = 0

        while password.count < length {
            if byteIdx >= expanded.count {
                // Need more bytes -- rehash (matches TS behavior)
                var moreInput = expanded
                moreInput.append(Data(String(password.count).utf8))
                let more = Data(HMAC<SHA512>.authenticationCode(for: moreInput, using: hmacKey))
                for i in 0..<more.count where password.count < length {
                    let charIdx = Int(more[i]) % charsetArray.count
                    password.append(charsetArray[charIdx])
                }
                break
            }

            let charIdx = Int(expanded[byteIdx]) % charsetArray.count
            password.append(charsetArray[charIdx])
            byteIdx += 1
        }

        return String(password)
    }
}

// MARK: - CommonCrypto Bridge

import CommonCrypto

// MARK: - String Helpers

extension String {
    func leftPadded(toLength length: Int, with character: Character) -> String {
        let deficit = length - count
        if deficit <= 0 { return self }
        return String(repeating: character, count: deficit) + self
    }
}
